### PR TITLE
Fix HTTP -> HTTPS CORS

### PR DESCRIPTION
**Pull request by Benjamin-Dobell**
When we're redirecting from HTTP -> HTTPS the redirect is handled by Nginx itself, our applications never see the original request. In this case we should return a CORS-compatible redirect, without doing so browsers block the redirect and our application never gets the expected HTTPS request. 

There's no security implications here as we're simply instructing the browser to use HTTPS. When the browser proceeds to fire the request over HTTPS, then the application can handle the request and chose whether or not to include CORS headers allowing the browser to handle the response.

### CODE REVIEW

**Comment by josegonzalez** [comment]
This seems like your application shouldn't be making ajax requests to non-http. Rather, it sounds like you have bad code so doing this is just a hack for enabling that.

Can you give me a real-world use case for this?

**Comment by Benjamin-Dobell** [comment]
Sorry, I renamed the issue to remove mention to preflights. The underlying issue really has little to do with `POST`/preflight requests, although they are affected as well.

However, browsers will even go so far as to block a browser's access to the _response_ of cross-origin `GET` request, if the response is missing a "Access-Control-Allow-Origin". This makes perfect sense as it prevents a malicious website `GET`ting bank account details from a logged in users etc. There is no pre-flight sent as `GET` requests don't have side effects.

Anyway, in our case we have clients accessing _public_ APIs and files over HTTP, this is of course legacy, we've had HTTP for years. However, we've only started mandating HTTPS (i.e. redirecting to HTTPS just last week).

When we made the switch forcing HTTPS we found that we'd broken clients websites because browsers weren't following the redirect; because the CORS headers were missing.

**Comment by josegonzalez** [comment]
So it sounds like there *is* a security risk introduced by doing this. Not all endpoints should be available for any domain for any given application. Apis are one place where you'd want this, but would you want this for a login form?

I think this is better handled by documenting this in our nginx docs as opposed to making this the default.

**Comment by Benjamin-Dobell** [comment]
The access header is only attached to the 301 redirect, to my knowledge there are zero security implications.

Browsers will fire off the request again over HTTPS as instructed and the application will then have a chance to decide whether or not it wants CORS per request.

The original CORS spec actually blocked this behaviour for preflight requests i.e. Old browsers will refuse to follow the 301 even though it makes no sense. That's since be removed from the spec as it made no sense - browsers should not just resubmit the preflight over HTTPS as instructed.

So basically _relying_ on this behaviour for any request subject to a preflight isn't a great idea. There's no security risk, it just may not work due to old browsers. However, for `GET` requests e.g. access to public APIs etc. then is should work flawlessly.

**Comment by prabirshrestha** [comment]
I would also be against this change. It is client responsibility to use https all the time. Given that lets encrypt makes it easy to get certs don't think it would be a problem.

Currently some browsers already block mixed contents.

**Comment by Benjamin-Dobell** [comment]
> I would also be against this change. It is client responsibility to use https all the time. Given that lets encrypt makes it easy to get certs don't think it would be a problem.
> 
> Currently some browsers already block mixed contents.

I think you may have misinterpreted. This pull request is specifically facilitating _using_ SSL. Without this pull request, third-parties that consume an API via over HTTP via CORS, can't be upgrade to SSL. You're stuck offering HTTPS, which is *bad*!

Yes, ideally you'd tell your third-parties to use HTTPS all the time - but ah, good luck on that. Particularly if we're talking old statically generated websites that are in the wild and unmaintained. Better to at least provide the response with SSL, rather than be forced to respond without any encryption.

**Comment by Benjamin-Dobell** [comment]
> Currently some browsers already block mixed contents.

Just to clarify. This pull requests facilitates *upgrading* HTTP -> HTTPS, not downgrading from HTTPS -> HTTP.

Mixed content is blocked when you access a website via HTTPS, then try load HTTP content. We're talking about the opposite

**Comment by josegonzalez** [comment]
I'm not going to merge this in. My reasoning is that I wouldn't personally want to have this behavior done automatically for my applications without knowing it is occurring.

Users that want this behavior can ship a custom nginx template. While that template does change from time to time, it should be stable enough to ship with the application.
